src/rust-2024/rpit-lifetime-capture.md

```diff
@@ -1,12 +1,313 @@
+
+```rust
+# #![feature(precise_capturing)]
+fn capture<'a, T>(x: &'a (), y: T) -> impl Sized + use<'a, T> {
+    //                                ~~~~~~~~~~~~~~~~~~~~~~~
+    //                             This is the RPIT opaque type.
+    //
+    //                                It captures `'a` and `T`.
+    (x, y)
+  //~~~~~~
+  // The hidden type is: `(&'a (), T)`.
+  //
+  // This type can use `'a` and `T` because they were captured.
+}
+```
+
+Currently, `use<..>` bounds must list all generic type and const parameters in scope, but they may exclude in-scope lifetime parameters.  E.g.:
+
+```rust
+# #![feature(precise_capturing)]
+fn capture<'a, T>(_: &'a (), _: T) -> impl Sized + use<T> {
+    //                                ~~~~~~~~~~~~~~~~~~~
+    //                 The opaque type captures only `T` and not `'a`.
+    ()
+}
+```
+
+You might want to capture more lifetimes in order to return a hidden type that uses those lifetimes.  Conversely, you might want to capture fewer lifetimes in order to allow the opaque type to be used in more places.  E.g., this is an error because the lifetime is captured despite the fact that the hidden type does not use the lifetime:
+
+```rust,compile_fail
+# #![feature(precise_capturing)]
+fn capture<'a>(_: &'a ()) -> impl Sized + use<'a> {}
+
+fn test<'a>(x: &'a ()) -> impl Sized + 'static {
+    capture(x)
+    //~^ ERROR lifetime may not live long enough
+}
+```
+
+Conversely, this is OK since the opaque type does not capture the lifetime:
+
+```rust
+# #![feature(precise_capturing)]
+fn capture<'a>(_: &'a ()) -> impl Sized + use<> {}
+
+fn test<'a>(x: &'a ()) -> impl Sized + 'static {
+    capture(x) //~ OK
+}
+```
+
+The bounds list of an RPIT opaque type may contain at most one `use<..>` bound.  If this bound is present, then the listed generic parameters are the only ones captured.
+
+### Edition-specific rules for elided `use<..>` bounds
+
+If the `use<..>` bound is *elided* (i.e., not present), then the compiler automatically inserts one according to edition-specific rules.
+
+In all editions, all in-scope type and const generic parameters are added to this automatically-inserted `use<..>` bound.  E.g.:
+
+```rust
+# #![feature(precise_capturing)]
+fn f_elided<T, const C: usize>() -> impl Sized {}
+//                                  ~~~~~~~~~~
+//                       The `use<..>` bound is elided here.
+//
+// In all editions, the above is equivalent to:
+fn f_explicit<T, const C: usize>() -> impl Sized + use<T, C> {}
+```
+
+In Rust 2021 and earlier editions, generic lifetime parameters are *not* included in this automatically-inserted `use<..>` bound for RPIT opaque types in the signature of bare functions and associated functions and methods within inherent impls.  However, starting in Rust 2024, these in-scope generic lifetime parameters *are* included.  E.g.:
```

Comment:
So first, I think we can shorten this by just saying "are *not* always captured". Also, I may have missed it, but do we actually explain *when* lifetimes are captured pre-2024?

Reply:
This paragraph has been changed, as elsewhere, to remove discussion of automatically inserted `use<..>` bounds, and it now states how lifetime parameters are captured in Rust 2021 and earlier editions.